test/hotspot/jtreg/compiler/lib/compile_framework/CompileFramework.java

```diff
@@ -0,0 +1,254 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.lib.compile_framework;
+
+import compiler.lib.compile_framework.SourceCode;
+import compiler.lib.compile_framework.CompileFrameworkException;
+import compiler.lib.compile_framework.InternalCompileFrameworkException;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URI;
+import java.net.URLClassLoader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import jdk.test.lib.process.ProcessTools;
+
+public class CompileFramework {
+    static final int JASM_COMPILE_TIMEOUT = 60;
+
+    private List<SourceCode> sourceCodes = new ArrayList<SourceCode>();
+    private URLClassLoader classLoader;
+
+    public void add(SourceCode sourceCode) {
+        sourceCodes.add(sourceCode);
+    }
+
+    public void printSourceCodes() {
+        for (SourceCode sourceCode : sourceCodes) {
+            System.out.println("SourceCode: " + sourceCode.className + " " + sourceCode.kind.name().toLowerCase());
+            System.out.println(sourceCode.code);
+        }
+    }
+
+    public void compile() {
+        if (classLoader != null) {
+            throw new RuntimeException("Cannot compile twice!");
+        }
+
+        printSourceCodes();
+
+        List<SourceCode> javaSources = new ArrayList<SourceCode>();
+        List<SourceCode> jasmSources = new ArrayList<SourceCode>();
+        for (SourceCode sourceCode : sourceCodes) {
+            switch (sourceCode.kind) {
+                case SourceCode.Kind.JASM -> { jasmSources.add(sourceCode);  }
+                case SourceCode.Kind.JAVA -> { javaSources.add(sourceCode);  }
+            }
+        }
+
+        String sourceDir = getSourceDirName();
+        compileJasmSources(sourceDir, jasmSources);
+        compileJavaSources(sourceDir, javaSources);
+        setUpClassLoader();
+    }
+
+    private static String getSourceDirName() {
+        // Create temporary directory for jasm source files
+        final String sourceDir;
+        try {
+            sourceDir = "compile-framework-sources-" + ProcessTools.getProcessId();
```

Comment:
Make it MT-safe, say by utilising the `File.createTempFile(prefix, suffix, directory)`?

Reply:
I would like the generated and saved files to be available int the `JTWork/scratch` directory. This allows us to download the source-code of a reproducer from our CI pipeline when there is a failure. Is that possible with `File.createTempFile` as well?

Reply:
But what exactly could go wrong here? The "compile" method is to be called in a single-threaded way. This thread should have a unique `ProcessTools.getProcessId`, right? So no two processes should be generating the same directory.

Reply:
>  Is that possible with File.createTempFile as well?
Yes. 'Directory', the argument is the path to contain the newly created temp file;

> But what exactly could go wrong here? The "compile" method is to be called in a single-threaded way.
1. Well, I personally don't like imposing such a heavy (single-threaded-only) requirement while we can easily make it mt-safe; It's a **universal** compiling FW, after all.
2. Among the most obvious users of the FW are automated test generators, the likes of JavaFuzzer and JITTester. The former is multi-threaded (and is loved in the industry for that). The latter presents serious problems (huge run times) because of its single-threaded nature;

Reply:
@lepestock Why exactly is my solution not multi-threading safe?
Does your solution with `File.createTempFile` allow (easy) extraction of the generated files when there is a failure? Where could those files be found?

Reply:
BTW: in the IR-Framework, we also use `ProcessTools.getProcessId` to generate unique filenames. Would that also be an issue there? So far we have not seen problems with it though...

Reply:
Thanks for the offline conversation. `ProcessTools.getProcessId` is per process. If there are multiple threads in one process, this leads to issues. Also if I use the `CompileFramework` repeatedly, this could lead to issues.

I think I can use `Files.createTempDirectory(Paths.get("."), "compile-framework-sources-");` to create a directory. "temp" does not mean it is necessarily deleted, only if I also use `File.deleteOnExit()`. So that should be ok, and the files are accessible if there is a failure.

I'm also thinking that I may need a temporary class-file directory. Otherwise, there could be issues if the `CompileFramework` is used repeatedly or in parallel. Currently, I just put them all in the `-Dtest.classes` dir, but that could lead to raise conditions.